### DISCUSSION TITLE
thetvdb API v4 to be monetized -- Medusa strategy?

### DISCUSSION THREAD

**Discussion by mmhere** [pymedusa/Medusa#8738] [General]
I guess this is either a feature request or a forward looking support question designed to track thetvdb API changes and impending monetization on their (thetvdb) part.

I just received an email that "v4" API represents their plan to monetize their data.  Projects either have to pay for a project apikey or pass the cost onto end users who will require their own personal apikey (global across projects they use).

I think I'm understanding it properly, but the email is attached below so it can speak for itself.

Deadline:  current APIs go offline in March 2021 according to the email.

---------

Apologies in advance for PDF here;  couldn't figure out how to reasonably preserve formatting and keep it "portable" (so-called).  Didn't want to paste the large block of text here either:

[x.pdf](https://github.com/pymedusa/Medusa/files/5541501/x.pdf)

**Comment by p0psicles** [comment]
Yeah I tried getting in contact, no success yet

**Comment by mmhere** [comment]
> Yeah I tried getting in contact, no success yet

Cool.  Thanks for initiating.

We could have a policy discussion separate from any comms happening with thetvdb.

I'd guess it unlikely that Medusa itself would be able to (or want to) pay for a license for the project.  Who knows, however, whether thetvdb might offer a free apikey at the Medusa project level?  Seems unlikely.

My conclusion, therefore, is that we need to [a] implement against their "v4" API that is supposedly live now, and [b] put in place user-level apikey capture and pass-thru for those that choose to pay individually.

I'm sure this will all go just as smoothly as thetvdb forced-turmoil that began October last yer.  ;-)

**Comment by p0psicles** [comment]
For me it means a. Make sure the alternatives work properly (tvmaze and today) and then b. See what we can do with tvdb changes

**Comment by DAVIZINH0** [comment]
Its posible a workarround to migrate all the series that actually i have with tvdb to the tvmaze indexer?
thanks

**Comment by OmgImAlexis** [comment]
If need be I can setup a proxy for us to use so they only see my key.

**Comment by p0psicles** [comment]
Tnx that's very generous. But, the issue with that is, that then we're getting into a cat&mouse game. Where it's possible they'll start blocking keys. Or IP addresses. And I'd like medusa to be an as much as possible stable app. 

So we need to work something out that's stable. Or we'll look into alternatives

**Comment by OmgImAlexis** [comment]
I’ve messaged them and sent them my contact info and nothing. They’re expecting $6k+ from another group. I feel like no matter what happens we’re not going to be in their good books. 💁‍♀️

**Comment by p0psicles** [comment]
Then we pull the plug when the time comes

**Comment by desepticon** [comment]
I guess one feature that would be nice to implement in the meantime is to be able to change the data provider. Either through a mass update or on a per-episode basis. I didn't see any option for that in the "edit" tab.

**Comment by p0psicles** [comment]
We just refactored add existing show page. I'll use  that to create an indexer change page. It's currently in develop branch if you like a preview

**Comment by desepticon** [comment]
> We just refactored add existing show page. I'll use that to create an indexer change page. It's currently in develop branch if you like a preview

 Thank!

**Comment by mmhere** [comment]
Do folks think keeping thetvdb as an option (with users putting in their own apikey, whether they get that by paying money for it, or by being gifted an apikey because they add data to thetvdb sufficiently and frequently enough[*]) is at all a reasonable option?

Or are we dumping them cuz their corporate takeover seeks to make money off data (which they don't even own) enter freely by users (who they don't even compensate).  (*cough*)  I think I risk starting a complaint fest here, which is not my intent.

If it is not too much additional development work here to track their "v4" API changes (sure to be ongoing, capricious, and arbitrary [and broken frequently]), it seems like it would be nice to keep access to their data as an option.  Aside from the data access methods and policy, the data when retrievable is pretty complete and usually of high quality.

Along the lines of which provider to use -- how do people feel about tvmaze?  Is it as fully populated with good data as is thetvdb?  Is it updated frequently enough?  Is its API access any more reliable?  Is the API itself any more stable? Given the many recent downtime problems with thetvdb and API breakages [bad/incomplete data served] over the past year, perhaps tvmaze is a better option overall if it would actually reliably serve [some] data?

I'm willing to switch entirely to tvmaze if it is up to snuff.  This is, however, a major undertaking given that show meta data is used by other platforms (various players) as well as here with Medusa.  I'd like to keep the retrievers and players in synch.  If changed in Medusa the effort to also change in players is painful.

[*]The communique I quoted above in PDF claims that thetvdb will have a policy (not yet fully defined apparently) that allows users who submit data to be given a valid apikey.  It sounds like this is one way to keep access to their data active without having to pay for it using legal tender, but instead pay for it with a sort of work trade.

**Comment by medariox** [comment]
We haven't decided yet. As a FOSS project the only thing that would make sense is to drop support completely. Their behavior in this situation has been despicable at best.

We have added support for TVmaze and TMDB since a long time and so far haven't received any complaints about either of them. Why don't you try them out with a few shows?

**Comment by mmhere** [comment]
Yes, I've started using TVmaze so it is good to hear that people haven't had issue with it.

Switching everything over is, as mentioned, a significant effort.

Along those lines it is very cool to see the migrate logic being developed (mentioned above) which will mitigate the pain of changing everything in Medusa.

Migrating in other [unnamed] projects/players poses a different challenge

---

> Their behavior in this situation has been despicable at best.

Agreed.  Keeping it FOSS (meaning use of TVmaze) is certainly preferred.

**Comment by mmhere** [comment]
Do we feel that the despicable behaviour precludes continuing to provide support for thetvdb?

Are there other examples of info providers (be it show meta data or other stuff, I dunno, subtitles) who force access thru non-FOSS mechanisms, which we still support?

**Comment by p0psicles** [comment]
For subtitle providers we do support supplying an api key to use them. But those are not part of Medusa core functionality. One thing to consider before dropping tvdb support is that anime users cannot make use of alternatives like tvmaze and tmdb, because thexem.de currently does not use them. Well, there are also some non-anime shows that depend on thexem.de. So that's something we should look out for. 

But I'd prefer if tvmaze or tmdb would add support for alternative season/episode numbering. So that a user can pick it when adding a show. Like select from indexer default, Netflix, scene. But thats more like a personal wish

**Comment by mmhere** [comment]
As far as implementing changes here to track the presumed-to-be moving target that is the "v4" thetvdb API (assuming we don't drop support), I'm willing to QA it.  I'm a SWdev IRL so I think I could be useful.  I plan to get a user apikey whether paid for by work trade or cash.

How likely are we to actually target the "v4" API then?  Anyone have a feel for this yet?  I certainly don't want to create extra dev work for functionality few will use if that is the case.

It does, however, seem quite unlikely that we would implement a project-level apikey for Medusa:

1. Seems highly unlikely they'd simply give this to us.
2. We surely aren't going to pay thousands of duckets for it as a project, right?

**Comment by p0psicles** [comment]
@thezoggy what are your thoughts on this?

**Comment by p0psicles** [comment]
Also started a discussion here. Tvmaze is very fast with feedback. https://www.tvmaze.com/threads/4515/alternative-episode-numbering#62122

@OmgImAlexis also inviting you to the discussion

**Comment by thezoggy** [comment]
TL;DR; It is still a little too early to tell. The new v4 api is not yet feature complete to replace v3 and with tvdb's sorrid history of api migrations I cant see it going smoothly or on time.

Personally, if they would have asked for donations on v1 days back before the v2+ mess I would have been fine to donate even after all the amount of time I've spent contributing to the site. The v2+ stuff turned me off. So for right now I would not recommend anyone giving them money until v4 is proven not to be a shitshow.

kodi/plex have giant user bases, they probably take up the bulk of tvdb api load. users of these type of apps dont really benefit tvdb as they never go to the website (so no chance to track user data / get ad views). but these apps have sponsors and other things that could help work out a deal and keep the user experience the same by not passing it down to the user. kodi for awhile now stopped defaulting to tvdb and moved over to tmdb for tvshows.. so might just see more people jump ship. also, someone would need to spend time on making a v4 compatible scraper for python2 (kodi 18 and older) and one for python3 (kodi 19)..  so that also adds some wrinkle onto who is going to spend time to do it.

sonarr for example has skyhook where all their users actually hit (skyhook gets data from tvdb, caches it and allows sonarr devs to clean up the data and do some logic on it). So really sonarr user base is reduced to minimal impact to tvdb because of this. also with them running the proxy that means they could do a key there but of course the one key for many is a no-no and why negotiations has to happen so it doesnt get abused. if you were to try and replicate this on the down low.. im sure your ip would get banned and so on.

**Comment by p0psicles** [comment]
@thezoggy tnx for the insight. Any plans on adding tvmaze or tmdb support to xem?

**Comment by thezoggy** [comment]
i had started to add tvmaze at one point (back when tvrage was around) but didnt because their data was so bad (they were pretty new of a site with low user base). i'll have to poke around again.

**Comment by nydave69** [comment]
No love for TheTVDB here.....

Just need an easy way to port over to tvmaze...

**Comment by Rouzax** [comment]
Just want to link my old feature request here 😄 
https://github.com/pymedusa/Medusa/issues/6531

**Comment by Frank071** [comment]
Just got an update for Synology Video Station mentioning the removal of the TVDB functionality because of the API change. That brought me here, since although I have SVS installed, I am using Medusa... The only reference to this upcoming TVDB change is this discussion thread that seems unfinished. I have looked for key elements mentioned in this discussion (like changing indexer, adding an api key, etc.) or any reference in the changelog, but found nothing of the sort. So my question is: what did I miss, can we help and if we are travelling into the direction of a termination of tvdb for medusa, should we then not mention it somewhere to avoid a lot of issues 'tvdb is no longer working'?

**Comment by Rouzax** [comment]
I recently switched my entire library from TVDB to TMDB and it was relatively smooth. Process I did was:

1. Delete shows from Medusa but keep files
2. Add existing show and added them again using TMDB.

To make sure my Kodi would still pick up everything I did manually create a tvshow.nfo in each show and only added the URL to TVDB and to TMDB. That way Kodi knows where to look with the scraper.
That results in a tvshow.nfo with only these lines:
```
https://thetvdb.com/series/a-very-english-scandal
https://www.themoviedb.org/tv/79299-a-very-english-scandal
```

Also create a Feature Request to get that added in Medusa to allow for an easier transition: 
https://github.com/pymedusa/Medusa/discussions/9134

**Comment by Frank071** [comment]
Well as it stands I currently do not even know what shows are linked to tvdb and which are not... And your (@Rouzax) approach probably works, but seems to be a rather hefty job given the number of shows I have on record....

_Edit: just checked that 'add existing show' feature (have not used it in the past) and that seems to be rather easy to do. Got some strange errors and it seems that it does not really pick up on already existing downloads automagically though. A tutorial might come in handy_

**Reply by p0psicles** [reply]
I think that as long the list is below 1000. It takes less time then you think. Also it a one time exercise.

Just start with a few to get a feel for the process

**Reply by Frank071** [reply]
Welll. I just tried it and although at first I thought it was not as difficult as I thought, it turned out it was... Some observations:
- (mass) removing the shows works, but a bit counter-intuitive ("delete" is positioned before "remove" and for me (non-native English) 'remove' also sounds as more destructive than delete) and lacks feedback (nothing seems to happen while in the background show after show is slowly removed from the grid.
- adding the existing show seems to completely ignore customised options (whatever the toggle setting is set to), so in my case it resulted in an overload of wanted and then snatched episodes.
- it seems that whilst an existing directory is added, the existing episode files are ignored, since already downloaded episodes get the default status setting (in stead of downloaded or archived)
- the adding of the shows has no feedback either, so shows will eventually pop up. Or not.
- the interface seems to suggest that one can select a couple of shows (via the checkboxes) and then add those (by pressing Start), but in fact the process then starts with the top X of the list where X is the number of boxes checked

Now all of this might seem a complaint rant, but that is not my intention. Given the fact that free access to the TVDB api is ending (when... it seems they even do not know themselves), I fear a lot of users with all kinds of backgrounds will need to migrate one way or the other. So even if there is no one-button-does-all, a tutorial describing what to do, what not to do and the caveats will really help in keeping the number of raised issues down....

(I will try to spend some time to turning some of my observations into proper issue descriptions, preferably with a hint of a cause)

**Reply by Rouzax** [reply]
Let me chime in my experience based on yours

> (mass) removing the shows works, but a bit counter-intuitive ("delete" is positioned before "remove" and for me (non-native English) 'remove' also sounds as more destructive than delete) and lacks feedback (nothing seems to happen while in the background show after show is slowly removed from the grid.

I agree this is a bit unclear and I also just did one show to see if I picked the right option

> adding the existing show seems to completely ignore customised options (whatever the toggle setting is set to), so in my case it resulted in an overload of wanted and then snatched episodes.

Do you mean the customized options that you set in the Add show page?
![image](https://user-images.githubusercontent.com/4103090/110752508-a8b94300-8245-11eb-9bfe-9d8fabad365a.png)

If you setup the scan like this and hit start, you can walk through the options per show and customize each one.
![image](https://user-images.githubusercontent.com/4103090/110752660-dbfbd200-8245-11eb-964b-b5e69b4f936e.png)
![image](https://user-images.githubusercontent.com/4103090/110752783-0188db80-8246-11eb-8a7c-14d2c083dffb.png)
![image](https://user-images.githubusercontent.com/4103090/110752899-241af480-8246-11eb-970c-c0f3328ca3de.png)


> it seems that whilst an existing directory is added, the existing episode files are ignored, since already downloaded episodes get the default status setting (in stead of downloaded or archived)

Those were picked up for me when I did a `Re-scan Files`

> the adding of the shows has no feedback either, so shows will eventually pop up. Or not.

The Add Existing show page is responsive and each added show will give you status. I think you did a big bang import of all existing shows at once. 

> the interface seems to suggest that one can select a couple of shows (via the checkboxes) and then add those (by pressing Start), but in fact the process then starts with the top X of the list where X is the number of boxes checked

Not sure about this one.

**Reply by Frank071** [reply]
Let's start with the important one:

> I think you did a big bang import of all existing shows at once.
Yes I did... since the interface seems to suggest that is possible if you want to apply the same (custom) options to all shows it find. So all my experiences are based on the big bang approache.

> > adding the existing show seems to completely ignore customised options (whatever the toggle setting is set to), so in my case it resulted in an overload of wanted and then snatched episodes.
> 
> Do you mean the customized options that you set in the Add show page?

Yes. And then I did not set that first toggle ("Prompt me...") but the second ("Use the customised ... for all shows"), since in my understanding this would apply all options set 'above' to all shows added 'below'

> If you setup the scan like this and hit start, you can walk through the options per show and customize each on
I did not try this show-by-show approach.

> > it seems that whilst an existing directory is added, the existing episode files are ignored, since already downloaded episodes get the default status setting (in stead of downloaded or archived)
> 
> Those were picked up for me when I did a `Re-scan Files`

Yes, that worked (to some extent) for me as well. The reason for mentioning it here is that this is not straightforward and probably also not what users would expect (since they add "an existing show"). For me the additional problem was that as the customised options were not picked up, all shows were added with the episodes marked 'wanted', which caused a storm of searches and downloads.

> > the adding of the shows has no feedback either, so shows will eventually pop up. Or not.
> 
> The Add Existing show page is responsive and each added show will give you status. I think you did a big bang import of all existing shows at once.

Here we disagree slightly. The page is not responsive, at least not with the big bang approach. If you add more than one show, all are marked "Running" but no progress is shown. Perhaps we have our wires crossed about 'responsive' ;-)

**Reply by p0psicles** [reply]
When a show is added you can hover over the add button. And it will show you the process. If you add existing using the checkboxes it will queue them all. And start adding shows one by one.

**Comment by mmhere** [comment]
Are we supporting the new tvdb API with access tokens (per user, 
presumably)?  Happy to test.

**Comment by HPMSGEN8** [comment]
What is the cost for the v4 API Per Year for medusa ?

**Reply by p0psicles** [reply]
We currently have no plans on updating tvdb api to a paid model.

**Reply by HPMSGEN8** [reply]
Hi @p0psicles 

If the community like myself would like to donate X to help fund this ? You and the team do so much to help keep the software alive and updated and safe so if the community like myself would like to give back then ? 

Just my 5 cents :)

**Reply by Rouzax** [reply]
This is however not a one time fee but one the comes back each year and that would mean the Devs need to beg each year to donate to be able to keep this feature.
The price is also quite steep from what I gathered.

**Reply by p0psicles** [reply]
rouzax has a good point. And we will never ever start paying for a service.
And regarding my personal oppionion on this. I'm not motivated at all to implement a paid service. Especially not if that service has been build up by thousands of user contributors.

I'm also not going to implement a service that has it's features locked by a paid model. The core functionality should be available to all users. With the exception of providers and notifications of course.

I also have to be honest that I don't know what the changes are for apiv4. And I don't know what the implications are for the apiv3. Like will they stop support for v3? No idea.

But tvdb has made clear which direction they are going in. And to me that means that i won't put a lot of effort into the indexer. I am however willing to put some more time into tvmaze and tmdb if that would be meaningful.

**Reply by Rouzax** [reply]
I'm now an avid user of TMDB because Trakt also moved that way.

**Reply by HPMSGEN8** [reply]
Hi @p0psicles 

I get your point now and agree :) Is there going to be an easy way to migrate shows between indexers ?

Thanks

**Reply by Rouzax** [reply]
You can use the methodology I described a bit earlier here.
Took me a bit of time, mostly because I also created an NFO file that pointed to the show URL on TMDB and tvdb, that way Kodi had no issues picking up the shows.

**Reply by p0psicles** [reply]
Only downside of tmdb is. They don't have airtime

**Comment by mmhere** [comment]
but please support the new API version for those who want to pay individually, yah?

On Monday 28 June 2021 12:33:10 PM (-07:00), p0ps wrote:



We currently have no plans on updating tvdb api to a paid model.

**Reply by p0psicles** [reply]
I don't even know that means? So can't say much about it.
But if it means that features from v3 api are going to be removed and moved to apiv4. Which would mean that the tvdb indexer would become unusable in medusa. Then i'm done with it.
But honestly I can't image any worthwile features that a paid apiv4 would offer, that I would be interested in adding to medusa.

**Comment by mmhere** [comment]
I don't think Medusa should pay for a project license, but I would like to be able to use API4 as an individual, paying my own fee.


but if it means Medusa can't reasonably _support_ API4, I get that.


at the moment, though, the tvmaze data (while pretty good) is still not *as* good as thetvdb data.


we'll have to see what thetvdb does to API3 as they move forward

On Tuesday 29 June 2021 11:21:19 AM (-07:00), p0ps wrote:



I don't even know that means? So can't say much about it.
But if it means that features from v3 api are going to be removed and moved to apiv4. Which would mean that the tvdb indexer would become unusable in medusa. Then i'm done with it.
But honestly I can't image any worthwile features that a paid apiv4 would offer, that I would be interested in adding to medusa.

**Reply by OmgImAlexis** [reply]
There's no good reason to move to the new paid API and members shouldn't be subscribing.

**Comment by mmhere** [comment]
There is because the data is still better at thetvdb.  Airtimes missing from others stops the searcher from finding episodes at the right time, for example.

On Thursday 01 July 2021 14:06:07 PM (-07:00), Alexis Tyler wrote:



There's no good reason to move to the new paid API and members shouldn't be subscribing.

**Reply by OmgImAlexis** [reply]
What others? Most of them can import from thetvdb so this isn't really an issue. Also from what I know the v3 isn't going away for quite a while and contains all the data we need.


Also please stop using emails to reply as this is creating a seperate top comment every time.

**Comment by ghost** [comment]
So unless I'm misunderstanding here, your response to thetvdb giving the middle finger to its users - the reason why it has the best data for shows by a mile - is to turn around and do the same to your users by purposely not supporting them even as a paid option, leaving users abandoned with no other options? (Sonarr isn't an option as it still has no idea how to handle backlogged episodes, and using anything other than thetvdb isn't an option either as the vast majority of the data I use/edit exists there and nowhere else.)

"Most of them can import from thetvdb so this isn't really an issue." 

I haven't seen this feature anywhere.

**Reply by p0psicles** [reply]
You can add shows to tvmaze as well as tmdb. How else are their shows sourced? 

And we don't give the middlefinger to our users. That's 100÷ on tvdb. Im just not motivated to put a lot of time in this, because tvdb got greedy. 

But if you want to sacrifice your hard earned free time, to add tvdb apiv4 support through a PR. I'll happily review it.

**Reply by LeeThompson** [reply]
I understand your view p0psicles, I just wish tvmaze didn't suck so much.

**Reply by Rouzax** [reply]
I've switched a few months ago to TMDB and so far have not run into any issues, KODI, Plex and Trakt still work without issue.
Did also switch in KODI to TMDB after upgrading to v19

**Comment by LeeThompson** [comment]
Can we at least have a quick way to change a show's indexer in Medusa without losing history?

**Reply by desepticon** [reply]
> Can we at least have a quick way to change a show's indexer in Medusa without losing history?

I also would love to have a feature like this. It would make this process so much easier if there was an indexer migration function in the mass update panel. And perhaps there should also be a notification for when the api finally goes offline as well, so users who might not be in the know can just click a button to transfer to a different one.

**Reply by p0psicles** [reply]
I'm working on a mass update indexer page.

**Reply by desepticon** [reply]
> I'm working on a mass update indexer page.

That would be very helpful. Thanks!

**Reply by p0psicles** [reply]
I could use some testers. You'd need to use git commands to checkout a specific branch.
any volunteers?

**Reply by p0psicles** [reply]
https://github.com/pymedusa/Medusa/pull/9862

**Comment by mmhere** [comment]
Since https://github.com/pymedusa/Medusa/issues/11217 and https://github.com/pymedusa/Medusa/issues/11218 mentioned this, I thought I'd post a fresh summary here so as not to pollute those other issues.

My interpretation of the discussion is...  correct me if my opinions colour this summary too much...

If not very many people were ever going to use the latest API [in Medusa] because it cost money, then there wasn't much motivation to implement and support client use of it in Medusa.  It would require effort to implement, effort to support, effort to bugfix, etc.

I was in favour of having support for it cuz I probly would pay, but I'm not a core developer and wouldn't want to burden them with having to do something that wouldn't get a whole lot of use.

That said, thetvdb is woefully inadequate at following thru on their "promises", so I'd be surprised to see them actually monetize the new API any time soon.  Would this make it worth trying to support it, um, temporarily?  Probly not because even if you get it running well on a temporary basis, if thetvdb ever does monetize, usage would fall way off.  This would still require support, and then be a burden for something with low usage.

**Comment by Minxster** [comment]
@medariox @mmhere ... As a newcomer to Medusa, from Sickchill (SC), I thought I'd just drop in a comment here. Nothing more than that though, just a comment with some musings 😉

I think we all knew, at some point, TVDB would end up causing massive problems to funnel people to their paid V4 model?!!! It's not the first and won't be the last time TVDB has caused upset. Though I can't imagine hosting such an API is cheap or without it's problems. But I do think this is them nailing the coffin for the free API. I expect many people who have been using SC to switch to Medusa (hopefully 😁👍) or something like Sonarr, since SC is still currently stuck with TVDB.

Will this now mean greater burdens on TVmaze, TMDB? I'm not counting IMDB since their API/solution/implementation seems a little more crazy than the others 🤣😂🤷‍♂️... I'm hoping that TVmaze + TMDB don't follow the same path as TVDB if their hosting costs drastically increase with people changing from TVDB v2🤷‍♂️

So... I think all of this needs another round of thought/ideas? I'm not saying these (below) are good ideas, but just trying to get those creative juices flowing 🤷‍♂️
- stick as is, with TVmaze and TMDB (and IMDB)
- work on the idea of something like Skyhook
- a 'go fun me' (type of thing) for Medusa to build a TVDB v4 plugin? Only in that, we all pay for our own API keys with TVDB, we just contribute to V4 plugin development?

Again, these are just thoughts and nothing more! Heck, I've not even finished my SC to Medusa migration yet! 🤣

**Reply by mmhere** [reply]
Skyhook?  Definition or reference appreciated.

**Reply by thezoggy** [reply]
> Skyhook? Definition or reference appreciated.

https://github.com/pymedusa/Medusa/discussions/8738#discussioncomment-215739
